data/data-pipeline/data_pipeline/comparison_tool/src/utils.py

```diff
@@ -0,0 +1,357 @@
+        population_df[geoid_column] / population_df[geoid_column].sum()
+    )
+
+    population_df["share_of_population_in_tracts"] = (
+        population_df[population_column]
+        / population_df[population_column].sum()
+    )
+
+    population_df.columns = [
+        "Population",
+        "Count of tracts",
+        "Share of tracts",
+        "Share of population",
+    ]
+    return population_df
+
+
+def get_demo_series(
+    grouping_column: str,
+    joined_df: pd.DataFrame,
+    demo_columns: list,
+) -> pd.DataFrame:
+    """Helper function to produce demographic information"""
+    # To preserve np.nan, we drop all nans
+    full_df = joined_df.dropna(subset=[grouping_column])
+    return (
+        full_df[full_df[grouping_column]][demo_columns]
+        .mean()
+        .T.rename(grouping_column)
+    )
+
+
+def get_tract_level_grouping(
+    joined_df: pd.DataFrame,
+    score_column: str,
+    comparator_column: str,
+    demo_columns: list,
+    keep_missing_values: bool = True,
+) -> pd.DataFrame:
+    """Function to produce segmented statistics (tract level)
+
+    Here, we are thinking about the following segments:
+    1. CEJST and comparator
+    2. Not CEJST and comparator
+    3. Not CEJST and not comparator
+    4. CEJST and not comparator
+
+    If "keep_missing_values" flag:
+    5. Missing from CEJST and comparator (this should never be true - it would be a tract we had not seen in CEJST!)
+    6. Missing from comparator and not highlighted by CEJST
+    7. Missing from comparator and highlighted by CEJST
+
+    This will make sure that comparisons are "apples to apples".
+    """
+    group_list = [score_column, comparator_column]
+    use_df = joined_df.copy()
+    if keep_missing_values:
+        use_df = use_df.fillna({score_column: "nan", comparator_column: "nan"})
+    grouping_df = use_df.groupby(group_list)[demo_columns].mean().reset_index()
+    # this will work whether or not there are "nans" present
+    grouping_df[score_column] = grouping_df[score_column].map(
+        {
+            True: "CEJST",
+            False: "Not CEJST",
+            "nan": "No CEJST classification",
+        }
+    )
+    grouping_df[comparator_column] = grouping_df[comparator_column].map(
+        {
+            True: "Comparator",
+            False: "Not Comparator",
+            "nan": "No Comparator classification",
+        }
+    )
+    return grouping_df.set_index([score_column, comparator_column]).T
+
+
+def format_multi_index_for_excel(
+    df: pd.DataFrame, rename_str: str = "Variable"
+) -> pd.DataFrame:
+    """Helper function for multiindex printing"""
+    df = df.reset_index()
+    df.columns = [rename_str] + [
+        ", ".join(col_tuple).strip()
+        for col_tuple in df.columns[1:].to_flat_index()
+    ]
+    return df
+
+
+def get_final_summary_info(
+    population: pd.DataFrame,
+    comparator_file: str,
+    geoid_col: str,
+) -> tuple[pd.DataFrame, str]:
+    """
```

Comment:
nit: can you put a one-line summary here?

Reply:
oof sorry about that, yes!